x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenReader.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.nlp.tokenizers;
+
+import java.util.LinkedList;
+import java.util.Optional;
+import java.util.Queue;
+
+import static java.lang.Character.isDigit;
+import static java.lang.Character.isLetter;
+import static java.lang.Character.isWhitespace;
+
+/**
+ * This is a char sequence tokenizer that satisfies the regex bpe regex for gpt-2
+ *      "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)|\\s+"
+ *
+ * It is by no means thread safe can only can parse one sequence at a time
+ */
+public class BpeTokenReader {
+    private final CharSequence inputChars;
+    int count;
+    private final Queue<Integer> putBackChar = new LinkedList<>();
```

Comment:
`ArrayDeque` is supposed to be faster than `LinkedList`.  Every time an item is added or removed to a linked list it needs to create or destroy the link structures.  But an array backed structure can continually reuse the same backing array and just adjust the pointers to the start and end of what's in use.

Reply:
@droberts195 I actually tried an arraydeque and it was slower in my benchmark (no matter the preallocated array deque size). In typical usage, we only ever put back one char. I was surprised by this, but the benchmark doesn't lie.